### DISCUSSION TITLE
`IDataListSourceValueConverter` conversion methods never fire?

### DISCUSSION THREAD

**Discussion by s6admin** [leekelleher/umbraco-contentment#300] [Discussion]
Hi leekelleher, 

Been stuck on this for weeks so I'm hoping you can offer some insight. Currently on the 4.4.0 branch, but haven't had success in ANY branch so I don't believe its version-specific. 

Originally followed the documentation example to create a [TimeZoneInfo DataList Source](https://github.com/leekelleher/umbraco-contentment/blob/develop/docs/editors/data-list.md), but the **ConvertValue()** method is NEVER called. I can see it is referenced during the TryGetPropertyTypeConfiguration method, but it is never executed when working on applicable nodes. The GetItems() method is called, so I'm pretty sure that the class is wired up correctly. 

I assumed this would be where picked values could be manipulated before they are passed to the backoffice model. Am I completely wrong about that?

**Comment by leekelleher** [comment]
@s6admin Sorry to hear that this has been a blocker for the past few weeks! 😬

Can you clarify if the custom data source is implementing the `IDataListSourceValueConverter` interface?

The `ConvertValue()` method is definitely being called on my own custom data sources, as I'm using it on my own CMS implementations.

**Reply by s6admin** [reply]
Yes, everything interfaces IDataListSourceValueConverter.
Even tried using the TimeZoneDataSource example directly from the DataList documentation and that never triggers ConvertValue() either.
What code from the Umbraco core does the ConvertValue() method hook into? I've tried debugging local versions of both the Contentment project and Umbraco 10 and can't get any related breakpoints to trigger.

**Reply by s6admin** [reply]
I assume something in Umbraco eventually triggers it because the only usage I see in the Contentment source is:

https://github.com/leekelleher/umbraco-contentment/blob/1f9462e2c625cd38691bda2756e74b8860f00ee9/src/Umbraco.Community.Contentment/DataEditors/DataList/DataListValueConverter.cs#L164

And that just looks an assignment rather than an execution.
I must be missing something incredibly obvious...

**Reply by leekelleher** [reply]
The method reference is stored to be called later on, depending whether it's a single or multiple item.
e.g. https://github.com/leekelleher/umbraco-contentment/blob/1f9462e2c625cd38691bda2756e74b8860f00ee9/src/Umbraco.Community.Contentment/DataEditors/DataList/DataListValueConverter.cs#L76

Still seems odd that it's not being called for you. I appreciate that it's frustrating. 😬

**Reply by s6admin** [reply]
Hmm...I can tell that DataListValueConverter.ConvertIntermediateToObject (nor the other variations) aren't being called either. Is there a way Contentment would skip those methods or is that where the Umbraco pipeline is in control?

**Reply by leekelleher** [reply]
The only reason (I can think of) why it wouldn't be called (within Contentment's control) is that the raw database value isn't a `string` and/or can't be deserialized into an `IEnumerable<string>` - but then that should throw an exception, so would be in the error logs.

**Reply by leekelleher** [reply]
Outside of Contentment's control, could the Value converter be disabled? are the values being saved programmatically (e.g. not a serialized JSON string array)?

**Reply by s6admin** [reply]
The logs are pretty quiet.
I'm working in a vanilla U10 install from nuget and a local copy of Contentment 4.4.0 branch for debugging, so I'm trusting most things are in their default state (ie. "enabled"). 
The data should be saving natively...I just whipped up a quick Contentment DataList using the **.NET Timezones** datasource with an **Item Picker** editor to try out the TimezoneInfo class example. When I follow along in the debugger the DataListValueConverter.ConvertIntermediateToObject() never even starts. The GetItems() method is always reached (to populate the possible Timezone choices) so at least that part is alive.
I wish I had more code to show but there really isn't much else going on in the project unless something bad happened during one of the nuget installations.

**Reply by leekelleher** [reply]
I'll try to reproduce tomorrow. Just to check, is this v10.4.0 (or earlier)? I'm wondering if something has changed in latest V10 Umbraco. 🤔

To note, I've been running latest Contentment on v10.3.2 and just started a v11.1.0 project where custom data sources have been working as expected.

**Reply by s6admin** [reply]
That would be fantastic! 
I'll keep trying on this end and ideally find more helpful details for you. 
Using Umbraco 10.3.2 from CLI installation. 
Have a good night.

**Reply by leekelleher** [reply]
Hey @s6admin - I tried it out, and not what you'll want to hear, but I wasn't able to reproduce the issue, it all worked as expected.

Here was my steps...

1. Installed Umbraco 10.3.2 from the command line, using the [Package Script Writer](https://psw.codeshare.co.uk/) to generate the install script, ([full options linked here](https://psw.codeshare.co.uk/?InstallUmbracoTemplate=true&UmbracoTemplateVersion=10.3.2&Packages=Our.Umbraco.Community.Contentment%7C4.4.0&UserEmail=admin%40example.com&ProjectName=MyProject&CreateSolutionFile=true&SolutionName=MySolution&UseUnattendedInstall=true&DatabaseType=SQLite&UserPassword=1234567890&UserFriendlyName=Administrator&IncludeStarterKit=false&StarterKitPackage=clean&OnelinerOutput=true).)

```ps
dotnet new -i Umbraco.Templates::10.3.2 && dotnet new sln --name "MySolution" && dotnet new umbraco --force -n "MyProject" --friendly-name "Administrator" --email "admin@example.com" --password "1234567890" --development-database-type SQLite && dotnet sln add "MyProject" && dotnet add "MyProject" package Our.Umbraco.Community.Contentment --version 4.4.0 && dotnet run --project "MyProject"
```

That installed Umbraco and Contentment, with default settings - I didn't make any changes to the "appsettings.json", so things like ModelsBuilder were all default.

I checked the backoffice, all appears to be working fine.

2. On the file system, (Windows Explorer), I created a plain-text file in the root of the webapp, named it "TimeZoneDataSource.cs".
3. Opened it in Notepad++ (as I didn't want to bring Visual Studio into the mix), and pasted in [the C# code snippet for the custom data-source with models](https://github.com/leekelleher/umbraco-contentment/blob/develop/docs/editors/data-list.md#providing-custom-values-for-published-content-models).
4. On the command line, I exited the process, <kbd>CTRL+C</kbd>, and re-compiled the project, e.g. `dotnet run --project "MyProject"`.
5. `error CS0246: The type or namespace name 'IDataListSource' could not be found (are you missing a using directive or an assembly reference?)`
6. Ah yes, namespaces... added `using Umbraco.Community.Contentment.DataEditors;` to the C# file.
7. Re-compile.
8. ` error CS0246: The type or namespace name 'ConfigurationField' could not be found (are you missing a using directive or an assembly reference?)`
9. Doh! Added `using Umbraco.Cms.Core.PropertyEditors;`
10. Re-compile.
11. ` error CS0535: 'TimeZoneDataSource' does not implement interface member 'IContentmentEditorItem.Group'`
12. Gah! 😫 My documentation is out of date! 🤦‍♀️🤦‍♂️
13. Added the Group property, e.g. `public string Group => "AAA";`
14. Re-compile.
15. Success!
16. Reloaded the backoffice, created new DataType for the Data List with the custom Time Zones data-source, with the Item Picker editor.
17. Created new DocType (with Template), with one property with an alias of "timeZones".
18. In the Content section, created new page, picked a few timezones, pressed Save & Publish.
19. Click on the Info tab, clicked on the page's URL, opened in new tab. Blank page because I hadn't edited the template yet.
20. Went to the template (within the backoffice), added the following...

```cshtml
<pre><code>@Model.TimeZones</code></pre>

<ul>
    @foreach (var zone in Model.TimeZones)
    {
        <li>@zone</li>
    }
</ul>
```

21. Reloaded the page on the frontend, and got the following...

<code>System.Collections.Generic.List`1[System.TimeZoneInfo]</code>

- (UTC+00:00) Dublin, Edinburgh, Lisbon, London

---

Sorry that I wasn't able to reproduce the issue you're experiencing. I'm wondering if any settings have been changed, maybe for ModelsBuilder? but then, I'm unsure what impact that has internally within Umbraco for the ValueConverters - I'd expect them to still be called.

**Reply by s6admin** [reply]
Still, this is great to see! Thank you for working through it. 
I will try to follow along and hopefully yield a different result than my current setup. 
The biggest issue is the value converters aren't firing when viewing the node in the backoffice like they did in U7/8 (I assume they still should in U9/10 ?). Being able to tweak the data when converting to/from the database is going to be critical for accomplishing Relation mapping as far as I can tell.

**Comment by enkelmedia** [comment]
I'll join in on this one as I had a similar experience and figured that what was wrong in my situation.

**I had the following scenario:**

I added the `IDataListSourceValueConverter`-interface to a custom IDataListSource in a Umbraco 11.2.1-project with Contentment 4.4.2. The breakpoint in `GetValueType`-method was hit during startup and my ModelBuilder-models was generated as I expect, even the values (output from ConvertValue) was provided as I expect when I looked at them during debugging (VS Quick Watch). 

But I never got any breakpoint-hits in the `ConvertValue`-method.

I did run the exact same code in the latest source code in the develop-branch and when running in that project the breakpoint did hit.

What's the difference!!? 

Turns out that I did add the property to a view in when testing in the source code-project. I went back to my other project and made sure to render the values from my property in a Razor-view. After rendering the value in a view the breakpoint starts to hit. The problem here is that when I was only looking at the values in the VS-debugger the breakpoint did not hit - but when rendering on the page the breakpoint is hit.

Not sure if this is the same underlying issue but figured I'll share my experience if it helps anyone.

Cheers!

**Comment by leekelleher** [comment]
Closing this off, I hope it was more of a quirk with Contentment v4.x - as it hasn't been raised again in the past few years.
